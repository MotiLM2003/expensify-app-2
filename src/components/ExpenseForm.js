import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    const { expense } = this.props;
    this.state = {
      id: expense ? expense.id : 0,
      description: expense ? expense.description : '',
      note: expense ? expense.note : '',
      amount: expense ? (expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: expense ? expense.calendarFocused : false,
      error: '',
    };
  }

  onDescriptionChange = (e) => {
    const desc = e.target.value;
    this.setState({ description: desc });
  };

  handleEvent = (e) => {
    const amount = e.target.value;
    if (
      (e.target.name === 'amount' &&
        (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))) ||
      e.target.name !== 'amount'
    ) {
      this.setState({ [e.target.name]: amount });
    }
  };

  onDateChange = (createdAt) => {
    if (!createdAt) {
      return;
    }
    this.setState(() => ({ createdAt }));
  };

  onFocusChanged = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onAddExpenseSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      // set error state.
      this.setState({
        error: 'Please provide description and amount :alien: ',
      });
    } else {
      this.setState({ error: '' });
      this.props.onSubmit({
        id: this.state.id,
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <React.Fragment>
        {this.state.error && (
          <p class='message warning mt-2 rad-xs'> {this.state.error}</p>
        )}
        <form
          onSubmit={this.onAddExpenseSubmit}
          className='expense-form-layout__form'
        >
          <input
            type='text'
            placeholder='Description'
            autoFocus
            name='description'
            value={this.state.description}
            onChange={this.handleEvent}
          />
          <input
            type='number'
            placeholder='Amount'
            name='amount'
            value={this.state.amount}
            onChange={this.handleEvent}
          />

          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={({ focused }) =>
              this.setState({ calendarFocused: focused })
            }
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder='Add a note for your expense (optinal)'
            name='note'
            value={this.state.note}
            onChange={this.handleEvent}
          ></textarea>

          <button className='button primary w-250 rad-xs'>Add expense</button>

          <div></div>
        </form>
      </React.Fragment>
    );
  }
}

export default connect()(ExpenseForm);
