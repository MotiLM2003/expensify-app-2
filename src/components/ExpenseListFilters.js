import React, { useState } from 'react';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = { cleandarFocused: null };
  onDateChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (x) => {
    this.setState({ cleandarFocused: x });
  };

  onSortChange = (e) => {
    const func =
      e.target.value === 'date'
        ? this.props.sortByDate
        : this.props.sortByAmount;
    func();
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  render() {
    return (
      <div>
        <input
          type='Text'
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDateChange}
          focusedInput={this.state.cleandarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

// export const ExpenseListFilters = (props) => {
//   const [cleandarFocused, setCleandarFocused] = useState(null);
// };
const mapState = (state) => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate()),
  };
};

export default connect(mapState, mapDispatchToProps)(ExpenseListFilters);
