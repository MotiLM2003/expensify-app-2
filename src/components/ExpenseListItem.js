import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';

export const ExpenseListItem = (props) => {
  const { id, description, amount, createdAt } = props;
  return (
    <div>
      <Link to={`/edit/${id}`}>
        {' '}
        <h1>{description}</h1>
      </Link>
      <p>
        id = {id}
        {amount} - create at: {moment(createdAt).format('DD/MM/yyyy')}
      </p>
    </div>
  );
};

export default connect()(ExpenseListItem);
