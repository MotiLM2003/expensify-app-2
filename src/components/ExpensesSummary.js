import React from 'react';
import { connect } from 'react-redux';
import { getVisibleExpenses } from '../selectors/expenses';
import { getExpensesTotal } from '../selectors/expense-total';
import numeral from 'numeral';

export const ExpensesSummary = ({ totalAmount, length }) => {
  const expenseWord = length === 1 ? 'expense' : 'expenses';
  const summary =
    length === 0 ? (
      'No expense available.'
    ) : (
      <React.Fragment>
        Viewing {length} {expenseWord} totaling
        {numeral(totalAmount).format('$0,0.00')}
      </React.Fragment>
    );
  return (
    <div>
      <h1>{summary}</h1>
    </div>
  );
};

const mapStateToProps = (state) => {
  const expenses = getVisibleExpenses(state.expenses, state.filters);
  return {
    totalAmount: getExpensesTotal(expenses),
    length: expenses.length,
  };
};
export default connect(mapStateToProps)(ExpensesSummary);
