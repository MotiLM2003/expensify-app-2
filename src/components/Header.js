import React from 'react';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import { startLogout } from '../actions/auth';

const Header = (props) => (
  <div>
    <h1>My App</h1>
    <ul className='nav'>
      <li>
        <NavLink to='/dashboard' exact activeClassName='active-link'>
          Dashboard
        </NavLink>
      </li>
      <li>
        <NavLink to='/create' activeClassName='active-link'>
          Create
        </NavLink>
      </li>

      <li>
        <NavLink to='/expense/' activeClassName='active-link'>
          Expense
        </NavLink>
      </li>
    </ul>
    <button
      onClick={() => {
        props.startLogout();
      }}
    >
      Log out
    </button>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: dispatch(startLogout),
});

export default connect(undefined, mapDispatchToProps)(Header);
