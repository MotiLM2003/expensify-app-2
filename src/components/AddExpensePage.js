import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <section className='expense-form-layout'>
        <h1>Add Expense Page</h1>
        <div className='container'>
          <ExpenseForm onSubmit={this.onSubmit} />
        </div>
      </section>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    startAddExpense: (expense) => dispatch(startAddExpense(expense)),
  };
};
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
