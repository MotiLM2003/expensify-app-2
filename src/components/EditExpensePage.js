import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onUpdateExpense = (expense) => {
    this.props.editExpense(expense);
    this.props.history.push('/dashboard');
  };

  onRemoveExpense = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <section className='expense-form-layout'>
        <h1>Edit Expense</h1>
        <div className='container'>
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onUpdateExpense}
          />
          <button
            className='button danger w-250 rad-xs'
            onClick={this.onRemoveExpense}
          >
            Remove expense
          </button>
        </div>
      </section>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    editExpense: (expense) => dispatch(editExpense(expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
  };
};

const stateToProps = (state, props) => () => {
  return {
    expense: state.expenses.find((x) => x.id === props.match.params.id),
  };
};

export default connect(stateToProps, mapDispatchToProps)(EditExpensePage);
