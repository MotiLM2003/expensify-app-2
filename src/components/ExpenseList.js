import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import { startGetExpenses } from './../actions/expenses';
import { getVisibleExpenses } from '../selectors/expenses';

export const ExpenseList = ({ expenses, getExpenses }) => {
  useEffect(() => {
    if (expenses.length === 0) {
      getExpenses();
    }
  }, [getExpenses]);
  return (
    <div>
      {expenses.length === 0 ? (
        <div>no Expense</div>
      ) : (
        expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense} />;
        })
      )}
    </div>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    getExpenses: () => dispatch(startGetExpenses()),
  };
};

const mapStateToProps = (state) => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseList);
