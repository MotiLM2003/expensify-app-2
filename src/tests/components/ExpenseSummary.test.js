import React from 'react';
import { expenses } from '../fixtures/expense';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
let wrapper,
  totalAmount = 100,
  length = 0;

beforeEach(() => {
  wrapper = shallow(<ExpensesSummary totalAmount={1000} length={0} />);
});

test('should render without expense', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render single expense', () => {
  wrapper.setProps({ totalAmount: 2000, length: 1 });
  expect(wrapper).toMatchSnapshot();
});



test('should render multiple expenses', () => {
    wrapper.setProps({ totalAmount: 2000, length: 3 });
    expect(wrapper).toMatchSnapshot();
  });
  