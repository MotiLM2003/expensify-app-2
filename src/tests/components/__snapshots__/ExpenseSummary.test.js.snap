// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totaling
    $2,000.00
  </h1>
</div>
`;

exports[`should render single expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling
    $2,000.00
  </h1>
</div>
`;

exports[`should render without expense 1`] = `
<div>
  <h1>
    No expense available.
  </h1>
</div>
`;
